README.md
# Concourse pool boy

Observes the pools used by the *Concourse pool resource* and resets to unclaimed any lock that it determines **stale**.

Can be used from the command-line or from a Concourse pipeline.

We strongly suggest to use our [forked version of the Concourse pool-resource], since this enables the Pool Boy to query the ATC and determine if a lock is stale based on correct liveliness information about the job which acquired it. To enable the Pool Boy to query the ATC, see section "Enabling ATC querying".

On the other hand, if paired with the original Concourse pool resource, the only criterion it can use for staleness is a timeout, so it is possible for the Pool Boy to steal a lock out of a perfectly fine job and cause mayhem when the job that owned the lock attempts to unclaim it.

## Status

We have been using this software in production, periodically triggered by a Concourse pipeline, for a few months and it is stable to use.

We follow semver numbering. We suggest to pin to a specific commit and perform tests before upgrading.

See also the [CHANGELOG](CHANGELOG.md).

## Usage from the command-line

```text
Usage: pool_boy.py [OPTIONS] COMMAND [ARGS]...

Options:
  --verbose                Turn on verbose logging
  --repo TEXT              URL of the Concourse lock pool repo  [required]
  --pools TEXT             Comma-separated list of pool name and timeout pairs. The pair
                           items must be separated by ":", for example:
                           worker_pool:60,tester_pool:30. The timeout parameter defines
                           the number of minutes after which the lock is considered
                           stale.  [required]
  --help                   Show this message and exit.

Commands:
  clean   Cleans the pool.
  status  Reports the status of the pool.
```

## Usage from a Concourse pipeline

Something along the lines of

```YAML
resources:
  - name: schedule.time
    type: time
    source:
      interval: 10m <= adjust accordingly, more often doesn't make sense

jobs:
  - name: pool-boy
    build_logs_to_retain: 100
    max_in_flight: 1
    on_failure:
      put: NOTIFY_SOMEHOW
      params:
        message: "Splash! The pool boy fell into the pool. Rescue him!"
    plan:
      - get: schedule.time
        trigger: true
      - task: pool-boy
        config:
          platform: linux
          image_resource:
            type: docker-image
            source:
              repository: MYDOCKER/pool-boy
              tag: latest
          run:
            path: /usr/local/bin/pool_boy.sh
        params:
          CONCOURSE_BASE_URL: http://my-concourse.local
          CONCOURSE_USERNAME: ((THE_CONCOURSE_USER))
          CONCOURSE_PASSWORD: ((THE_USER_PASSWORD))
          POOL_REPO: MY_CONCOURSE_LOCKS_REPO
          POOL_REPO_SSH_PRIVATE_KEY: ((MY_GIT_SSH_PRIVATE_KEY))
          POOLS: POOL_1:STALE_TIMEOUT_1,POOL_2:STALE_TIMEOUT_2
```

## Enabling ATC querying

To enable the Pool Boy to query the ATC and determine if a lock is stale based on correct liveliness information about the job which acquired it:

1. Use our [forked version of the Concourse pool-resource].
2. Set `CONCOURSE_BASE_URL`, `CONCOURSE_USERNAME` and `CONCOURSE_PASSWORD` as follows.

`CONCOURSE_BASE_URL` is optional. It should be identical to the `--concourse-url` value passed to
`fly`. If present the `CONCOURSE_USERNAME` and `CONCOURSE_PASSWORD` are mandatory and they must be
valid Concourse credential with the appropriate rights to see the status of the builds that uses
the configured `POOL_REPO`.

## Caveats

Note that if you don't follow section "Enabling ATC querying", then the Pool Boy won't be able to validate if the build that last claimed a lock is still alive or not and will then rely solely on the staleness timeout configured.

This will typically result in a pipeline failure when the following sequence of events happen:

1. A job acquires lock `lock-1`.
2. Time passes by, the Pool Boy detects `lock-1` as stale and brings it back to the available pool.
3. The job of step 1 finishes running and attempts to release `lock-1`. The pool resource will fail (because the lock disappeared on git, moved by the Pool Boy) and the associated pipeline will fail!

So think twice before deciding the timeout values!

## Testing

The script works out-of-the box on macOS and Linux. You can optionally run it from a Docker container.

    docker build --tag MYORG/pool-boy:latest .

Make your private SSH key available to the Docker container:

This works if the host is Linux, but is broken if the host is Mac (https://github.com/docker/for-mac/issues/483)

    docker run \
        --volume $SSH_AUTH_SOCK:/ssh-agent \
        --env SSH_AUTH_SOCK=/ssh-agent \
        --interactive --tty --rm \
        MYORG/pool-boy:latest

Run the container and get a shell:

    docker run \
    --interactive --tty --rm MYORG/pool-boy:latest

Use local remote for quick feedback loop

One shot:

    git clone --bare MY_CONCOURSE_LOCKS_REPO concourse-pool-boy-test-local

Replace the repo URL with your own repo.

Then:

    pool_boy.py --repo (pwd)/concourse-pool-boy-test-local --pools testers status
    pool_boy.py --repo (pwd)/concourse-pool-boy-test-local --pools testers clean

## TODO

* Add Prometheus metrics via pushgateway.
* The final form should actually be a Concourse resource, instead than a script that is run as a pipeline task.

## License

Copyright 2018 Pix4D

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

### Vendor dependencies

* brigit by Kozea, BSD Licence
* additional dependencies listed in requirements.txt

    git subtree add -m 'Import github.com/Kozea/brigit.git 4fae4e5 2016-06-20' --squash --prefix vendor/brigit git://github.com/Kozea/brigit.git 4fae4e516ed77929037ba5d95eba2eca69faffbf

Note: there is no built-in help for git subtree, on the other hand the very good help is at https://github.com/git/git/blob/master/contrib/subtree/git-subtree.txt


[forked version of the Concourse pool-resource]: https://github.com/Pix4D/pool-resource


CHANGELOG.md
# Pool Boy Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2019-01-07

No changes; decided to promote to 1.0.0 thanks to its stability. More details in the documentation.

## [0.3.0] - 2018-12-13

### Added

- The pool boy can check, by querying the Concourse ATC, if a floating body is dead or alive before cleaning it from the pool.

## [0.2.0] - 2018-12-03

### Changed

- Made the timeout configurable per pool.

## [0.1.0] - 2018-10-17

### Added

- Say hello to the Pool Boy. Consumable as a Docker image for a Concourse task.


pool_boy.py
#!/usr/bin/env python3

"""
Clean up leaves and corpses from the pools.
"""

import glob
import logging
import os
import os.path
import re
import shutil
from datetime import datetime, timedelta, timezone

import click
import dateutil.parser
import requests
from vendor.brigit.brigit import Git

CONCOURSE_BASE_URL = os.environ.get('CONCOURSE_BASE_URL')
CLAIM_COMMIT_RE = re.compile(
    r'[0-9a-f]+\s+'
    r'(?P<team>.*)/(?P<pipeline>.*)/(?P<job>.*)\s+build\s+(?P<build>.*)\s+'
    r'claiming:\s+.*'
)

logging.basicConfig(format='%(message)s')
log = logging.getLogger('pool-boy')

DIRTY_POOLS_WORK_DIR = 'dirty-pools'
conf = {}


def refresh_local_repo(conf):
    os.makedirs(DIRTY_POOLS_WORK_DIR, exist_ok=True)

    os.chdir(DIRTY_POOLS_WORK_DIR)
    log.debug('Working directory %s', os.path.abspath('.'))

    local_repo = os.path.abspath(conf['local-repo'])
    if os.path.exists(local_repo):
        log.debug('Found working repo %s. Removing it', local_repo)
        shutil.rmtree(local_repo)
    else:
        log.debug('No working repo found')
    log.info('Cloning %s', conf['remote-repo'])
    # Doing a shallow clone sometimes takes even more than a full clone, so we keep things simple.
    git = Git(local_repo, conf['remote-repo'])

    log.debug('Setting committer info for the repo')
    git.config('user.name', 'Pool Boy')
    git.config('user.email', '<pool-boy@localhost>')

    return git


class BearerAuth(requests.auth.AuthBase):
    def __init__(self, token):
        self.token = token

    def __call__(self, request):
        request.headers['Authorization'] = 'Bearer ' + self.token
        return request


def get_concourse_auth():
    """Exchange the username and password for a token

    See: https://www.oauth.com/oauth2-servers/access-tokens/password-grant/
    """
    if CONCOURSE_BASE_URL is None:
        return None
    url = CONCOURSE_BASE_URL + '/sky/token'
    username = os.environ['CONCOURSE_USERNAME']
    password = os.environ['CONCOURSE_PASSWORD']
    reply = requests.post(
        url,
        auth=('fly', 'Zmx5'),  # See: https://github.com/concourse/concourse/blob/3792bda9c3705ce3a10f03d18b2768b9954fc51d/skymarshal/skyserver/skyserver.go#L298
        data=dict(username=username,
                  password=password,
                  grant_type='password',
                  scope='openid profile email federated:id groups')
    )
    if reply.ok:
        return BearerAuth(reply.json()['access_token'])
    log.error(f'Failed to authenticate as {username!r} at {url}')
    return None


def get_build_status(auth, team, pipeline, job, build):
    """Check with Concourse if a build is still alive

    The returned value is either:
    - None: If a meaningful reply from Concourse could not be obtained
            either due to networking issue or simply because the pipeline
            got destroyed.
    - The build status as reported by Concourse. Live builds are 'started'.
    """
    if not auth:
        return None

    reply = requests.get(
        CONCOURSE_BASE_URL + f'/api/v1/teams/{team}/pipelines/{pipeline}/jobs/{job}/builds/{build}',
        auth=auth
    )
    if not reply.ok:
        return None
    return reply.json()['status']


def clean_pool(git, pool, stale_timeout, dry_run):
    changes = 0
    log.info('\n====== Looking for claimed locks in pool %s (stale timeout %s) ======',
             pool, stale_timeout)
    assert os.path.isdir(pool), "Pool dir %s doesn't exist" % os.path.abspath(pool)
    claimed_dir = os.path.join(pool, 'claimed')
    unclaimed_dir = os.path.join(pool, 'unclaimed')

    def clear_lock(taken_lock):
        if not dry_run:
            free_lock = os.path.join(unclaimed_dir, os.path.basename(taken_lock))
            log.info(f"Moving {taken_lock} to {free_lock}")
            git.mv(taken_lock, free_lock)

    locks = glob.glob(os.path.join(claimed_dir, '*'))
    if not locks:
        log.info('No claimed locks')
        return changes
    now_ts = datetime.now(timezone.utc)
    auth = get_concourse_auth()
    for taken_lock in locks:
        log.info('---')
        lock_acquire_msg = git.log('--pretty=oneline', '--max-count=1', taken_lock).rstrip()
        log.info('lock: %s (%s)', taken_lock, lock_acquire_msg)
        match = CLAIM_COMMIT_RE.match(lock_acquire_msg)
        build_status = get_build_status(auth=auth, **match.groupdict()) if match else None
        # Format is ISO-8601: 2018-10-04T12:39:47+00:00
        claim_str = git.log('--format=%cI', '--max-count=1', taken_lock)
        claim_ts = dateutil.parser.parse(claim_str)
        log.debug('time now        %s', now_ts)
        log.debug('claim timestamp %s', claim_ts)
        lifetime = now_ts - claim_ts
        if build_status == 'started':
            log.info(f"Owning build is still alive after {lifetime}. Leaving lock as is.")
            continue
        elif build_status is not None:
            log.info(
                f"Owning build is terminated with status {build_status!r} (lifetime: {lifetime}).")
            changes += 1
            clear_lock(taken_lock)
        elif lifetime > stale_timeout:
            log.info(f"Couldn't check the build status and lock is stale (lifetime: {lifetime}).")
            changes += 1
            clear_lock(taken_lock)
        else:
            log.info(f"Couldn't check the build status "
                     f"but lock is not stale yet (lifetime: {lifetime}).")
    return changes


@click.group()
@click.option('--verbose', default=False, is_flag=True,
              help='Turn on verbose logging')
@click.option('--repo', required=True,
              help='URL of the Concourse lock pool repo')
@click.option('--pools', required=True,
              help='Comma-separated list of pool name and timeout pairs. '
                   'The pair items must be separated by ":", '
                   'for example: worker_pool:60,tester_pool:30. '
                   'The timeout parameter defines the number of minutes after which '
                   'the lock is considered stale.')
def cli(verbose, repo, pools):
    if verbose:
        log.setLevel(logging.DEBUG)
    else:
        log.setLevel(logging.INFO)
    local_repo = repo.split('/')[-1].split('.')[0]
    assert '/' not in local_repo
    conf['remote-repo'] = repo
    conf['local-repo'] = local_repo
    conf['pools'] = list(_parse_pools(pools))
    log.debug('Configuration: %s', conf)


def _parse_pools(pools):
    for part in pools.split(','):
        name_timeout = part.split(':', 1)
        assert len(name_timeout) == 2
        yield (name_timeout[0], timedelta(minutes=int(name_timeout[1])))


def clean_pools(git, pools, dry_run):
    return sum(clean_pool(git, name, timeout, dry_run) for (name, timeout) in pools)


@cli.command()
def status():
    """Reports the status of the pool."""
    git = refresh_local_repo(conf)
    os.chdir(conf['local-repo'])
    changes = clean_pools(git, conf['pools'], dry_run=True)
    log.info('\nSummary: detected %d changes', changes)


@cli.command()
def clean():
    """Cleans the pool."""
    git = refresh_local_repo(conf)
    os.chdir(conf['local-repo'])
    changes = clean_pools(git, conf['pools'], dry_run=False)
    if changes:
        log.info('\nChanges to commit:')
        log.info(git.status('--short'))
        log.info('Committing')
        git.commit('--all',
                   '--message=Freshening up the pool with %s changes' % changes)
        log.info('Pushing to %s', git.remote('get-url', '--push', 'origin'))
        git.push('origin', 'master')
    else:
        log.info('\nNo changes to push')


if __name__ == '__main__':
    cli()  # pylint: disable=no-value-for-parameter
    log.info('Done')
